engine/resources/pippet.go

```diff
@@ -0,0 +1,183 @@
+// Mgmt
+// Copyright (C) 2013-2018+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"fmt"
+	//"sync"
+
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/engine/traits"
+)
+
+func init() {
+	engine.RegisterResource("pippet", func() engine.Res { return &PippetRes{} })
+}
+
+// PippetRes is a wrapper resource for puppet.
+// It implements the functional equivalent of an exec resource
+// that calls "pupept resource <type> <title> <params>", but offers
+// superior performance through a long-running Puppet process that
+// receives resources through a pipe (hence the name).
+type PippetRes struct {
+	traits.Base // add the base methods without re-implementation
+	traits.Refreshable
+
+	init *engine.Init
+
+	Type   string `yaml:"type" json:"type"`
+	Title  string `yaml:"title" json:"title"`
+	Params string `yaml:"params" json:"params"`
+
+	runner PippetRunner
+}
+
+// Default returns an example Pippet resource.
+func (obj *PippetRes) Default() engine.Res {
+	return &PippetRes{
+		Params: "{}",
+	}
+}
+
+// Validate never errors out. We don't know the set of potential
+// types that Puppet supports. Resource names are arbitrary. We
+// cannot really validate the parameter YAML, because we cannot
+// assume that it can be unmarshalled into a map[string]string.
+func (obj *PippetRes) Validate() error {
+	return nil
+}
+
+// Init makes sure that the PippetRunner object is initialized.
+func (obj *PippetRes) Init(init *engine.Init) error {
+	obj.init = init // save for later
+
+	var err error
+	obj.runner, err = PippetRunnerInstance()
+	if err != nil {
+		return err
+	}
+
+	obj.runner.Register(obj)
+
+	return nil
+}
+
+// Close is run by the engine to clean up after the resource is done.
+func (obj *PippetRes) Close() error {
+	obj.runner.Unregister(obj)
```

Comment:
if err := ?

Reply:
I went for `return obj.runner.Unregister(obj)`. Is that fine?